### PR TITLE
Slurm nodelist parse simplify

### PR DESCRIPTION
**Pull request by dmageeLANL**
Simplify slurm nodelist parsing.  Uses only string methods.

### CODE REVIEW

**Comment by Paul-Ferrell** [comment]
There's a 'node_list_parsing' test you can run to check against all known regressions (apparently not including the one you found).
Run `./run_tests -q -o node_list_parsing` in the test directory.

Right off the bat I see a problem with your code. This is a valid nodelist: `foo003,foo0[10-20],foo[103-104],foo[10,12-14],foo-m11-16`. Not just any of those individual items, but the entire set of items.

**Comment by dmageeLANL** [comment]
New commit does the right thing, but doesn't have error checking.  I'm on the right path, I'll be done shortly.

File: lib/pavilion/schedulers/plugins/slurm.py [hunk 1]
```diff
@@ -186,6 +188,80 @@ def slurm_states(state):
...
+    for num in nums:
+        num2 = num.replace('-','')
+        if not num[-1].isdigit() or num2.isalnum():
+            char_error = num
+            break
+        if '-' in num:
+            numsplit = num.split('-')
+            if not numsplit[0][-1].isdigit():
+                char_error = num
+                break
```
Comment (Paul-Ferrell) [inline lib/pavilion/schedulers/plugins/slurm.py:234]: This check is redundant.

File: lib/pavilion/schedulers/plugins/slurm.py [hunk 2]
```diff
@@ -186,6 +188,80 @@ def slurm_states(state):
...
+            nodenums.extend([str(na) for na in nrng])
+        else:
+            nodenums.append(num)
+
+    # If the loop was broken for trailing nonsense. raise value error.
+    if char_error:
+        raise ValueError(
+            "Node spec element {} constains invalid characters."
+            .format(char_error)
+        )
```
Comment (Paul-Ferrell) [inline lib/pavilion/schedulers/plugins/slurm.py]: There's only one place you need to check for this, and the exception should be thrown there.

File: lib/pavilion/schedulers/plugins/slurm.py [hunk 3]
```diff
@@ -186,6 +188,80 @@ def slurm_states(state):
...
+    # Find the node number strings that are shorter than they should be.
+    # Collect their position in the node number list and the padded node number
+    # string.
+    replacenodes=[]
+    for i, anode in enumerate(nodenums):
+        lendiff = lennum - len(anode)
+        if lendiff > 0:
+            padname='0'*lendiff
+            newnode=padname+anode
+            replacenodes.append((i,newnode))
```
Comment (Paul-Ferrell) [inline lib/pavilion/schedulers/plugins/slurm.py:263]: Aw hell no. 
Use string formating: `'{prefix}{num:0{digits}d}'.format(prefix=prefix, num=num, digits=digits)`

File: lib/pavilion/schedulers/plugins/slurm.py [hunk 4]
```diff
@@ -313,72 +389,42 @@ def _get_config_elems(self):
...
+                    inner_collect = []
+            elif "[" not in node_section:
+                allnodes.append(node_section)
             else:
-                nodes.append(part)
+                if "]" in node_section:
+                    allnodes.append(node_section)
+                else:
+                    inner_collect.append(node_section)
+                    inner_split = True
```
Comment (Paul-Ferrell) [inline lib/pavilion/schedulers/plugins/slurm.py:426]: This is taking the Python way and trying to fix it. It's much cleaner and more efficient to do this the C way:

```python
 start = 0
depth = 0
node_list = []
for i, char in enumerate(test_data):
    if char == '[':
        if depth == 0:
            depth = 1
        else:
            raise ValueError("Extra open bracket in '{}'".format(test_data[start:i+1]))
    if char == ']':
        if depth != 1:
            raise ValueError("Extra close bracket in '{}'".format(test_data[start:i+1]))
        else:
            depth = 0
    if char == ',' and depth == 0:
        node_list.extend(parse_node_list_element(test_data[start:i])
        start = i+1 
```

**Review by Paul-Ferrell — CHANGES_REQUESTED** [review]

File: lib/pavilion/schedulers/plugins/slurm.py [hunk 5]
```diff
@@ -186,6 +188,80 @@ def slurm_states(state):
...
+        )
+
+    # Get the node prefix by splitting on the starting square bracket.
+    # Strip the end bracket and split into a list of strings on comma.
+    # If the string has a dash it's a range, otherwise it's a single node.
+    # Remember the length of the first node number string in the list.
+    # Make the numbers into integers, create ranges and collect them as strings.
+    isplit = node_spec.split("[")
+    prefix = isplit[0]
+    nums = isplit[-1].strip("]").split(',')
```
[the remark this answers is not in the document]
Comment (Paul-Ferrell) [reply lib/pavilion/schedulers/plugins/slurm.py:220]: On whitespace:
slurm accepts this: `fg0[11-12] , fg009`
but not this: `fg0[9 - 11]`

[the remark this answers is not in the document]
Comment (Paul-Ferrell) [reply lib/pavilion/schedulers/plugins/slurm.py:220]: Well, technically it accepts both, but it incorrectly calculates the value size when whitespace is involved. (It things `fg[9 - 11]` should be `fg09, fg011`

File: lib/pavilion/schedulers/plugins/slurm.py [hunk 6]
```diff
@@ -1,7 +1,9 @@
 # pylint: disable=too-many-lines
 """The Slurm Scheduler Plugin."""
 
+from curses.ascii import isalnum
 import math
+from multiprocessing import allow_connection_pickling
```
[the remark this answers is not in the document]
Comment (dmageeLANL) [reply lib/pavilion/schedulers/plugins/slurm.py:6]: Weird. I think VS code did this on it's own.

[the remark this answers is not in the document]
Comment (dmageeLANL) [reply lib/pavilion/schedulers/plugins/slurm.py:6]: I got it working though, I'll work through the comments tomorrow.

File: lib/pavilion/schedulers/plugins/slurm.py [hunk 7]
```diff
@@ -186,6 +188,80 @@ def slurm_states(state):
...
+        raise ValueError(
+            "Invalid Node List: '{}'. Unmatched range bracket []"
+            .format(node_spec)
+        )
+
+    # Get the node prefix by splitting on the starting square bracket.
+    # Strip the end bracket and split into a list of strings on comma.
+    # If the string has a dash it's a range, otherwise it's a single node.
+    # Remember the length of the first node number string in the list.
+    # Make the numbers into integers, create ranges and collect them as strings.
```
[the remark this answers is not in the document]
Comment (dmageeLANL) [reply lib/pavilion/schedulers/plugins/slurm.py:218]: It already is handled in the bracket_check=0 case.

File: lib/pavilion/schedulers/plugins/slurm.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (dmageeLANL) [reply lib/pavilion/schedulers/plugins/slurm.py:221]: I just removed all whitespace in the calling method.

[the remark this answers is not in the document]
Comment (dmageeLANL) [reply lib/pavilion/schedulers/plugins/slurm.py:221]: And it only accepts two square brackets per node list element now.

File: lib/pavilion/schedulers/plugins/slurm.py [hunk 8]
```diff
@@ -186,6 +188,80 @@ def slurm_states(state):
...
+    isplit = node_spec.split("[")
+    prefix = isplit[0]
+    nums = isplit[-1].strip("]").split(',')
+    nlims = nums[0].split('-')
```
[the remark this answers is not in the document]
Comment (dmageeLANL) [reply lib/pavilion/schedulers/plugins/slurm.py:222]: Ok.  I mean spaces are invalid in a node list, but their meaning is clear.  So get rid of them.

File: lib/pavilion/schedulers/plugins/slurm.py [hunk 9]
```diff
@@ -186,6 +188,80 @@ def slurm_states(state):
...
+    lennum = len(nlims[0])
+    nodenums = []
+    char_error = 0
+    for num in nums:
+        num2 = num.replace('-','')
+        if not num[-1].isdigit() or num2.isalnum():
```
[the remark this answers is not in the document]
Comment (dmageeLANL) [reply lib/pavilion/schedulers/plugins/slurm.py]: Excellent point.

File: lib/pavilion/schedulers/plugins/slurm.py (hunk 3 above)
Comment (dmageeLANL) [reply lib/pavilion/schedulers/plugins/slurm.py:264]: Que?

Comment (Paul-Ferrell) [reply lib/pavilion/schedulers/plugins/slurm.py:253]: You're assembling the string manually, when there are built-in mechanisms for that.

File: lib/pavilion/schedulers/plugins/slurm.py [hunk 10]
```diff
@@ -312,71 +384,42 @@ def _get_config_elems(self):
...
+                    allnodes.append(','.join(inner_collect))
+                    inner_collect = []
+            elif "[" not in node_section:
+                allnodes.append(node_section)
+            else:
+                if "]" in node_section:
+                    allnodes.append(node_section)
+                else:
+                    inner_collect.append(node_section)
+                    inner_split = True
```
Comment (Paul-Ferrell) [inline lib/pavilion/schedulers/plugins/slurm.py:415]: You're still doing what I advised against, though this time as a state machine.  I'm not going to waste too much time arguing about this, but I really think the character by character method is much more readable.  It also has the added benefit of being able to find and print errors about extra parens and bad white space in a way that can give the exact context where it occurred.

**Comment by Paul-Ferrell** [comment]
This got replaced by a standard library.

**Pull request closed without merging**